improve path helper methods

mjauvin opened the pull request:
- fix uploads_path() helper method
- add media_path() helper method

mjauvin committed:
- fix uploads_path() helper method
- add media_path() helper method

LukeTowers commented:
@mjauvin please post a description of how the existing uploads_path() helper is broken. Additionally, shouldn't both those paths support being configurable just like the themes & plugin paths are?

mjauvin commented:
The uploads_path() helper was referencing  app('path.uploads') which didn't exist...

As for configurability, since their value is taken from config, I though this would be redundant to also make it changeable in the Application class... what do you think @LukeTowers?

bennothommo commented:
@mjauvin Unfortunately, this change doesn't seem to take into account the `storage.uploads.path` and `storage.media.path` configuration settings.

mjauvin committed:
also make temp, uploads and media path configurable

mjauvin commented:
@bennothommo you mean from the Application class? I may have read that wrong, but it looks like the Application class holds the default values for those paths, which can be overriden in the cms config files... am I getting this wrong?

bennothommo commented:
@mjauvin Ah sorry, you're right. Just as long as the helpers are referencing and allowing the correct configuration, which it seems to be, that's all good.

mjauvin committed:
add the class path properties

mjauvin commented:
need to update the docs as well for `media_path()`

mjauvin commented:
https://github.com/octobercms/docs/pull/456

LukeTowers commented:
With `$this->basePath` being added, does this still work for remote storage drivers? Could you add some unit tests for these helpers?

mjauvin commented:
Not sure what you mean @LukeTowers, I only added methods which didn't exist in Application class and fixed the one that was broken in the helpers.

mjauvin commented:
But yeah, I can look at adding unit tests for the Application class.

mjauvin commented:
gotta get back to this, please keep opened

mjauvin committed:
add base application tests

mjauvin committed:
fix missing ";"

mjauvin committed:
fix code quality warnings

mjauvin committed:
add some support helpers tests

mjauvin committed:
actually use the app defined in setup

mjauvin committed:
add tests for Support/helpers

mjauvin committed:
remove extra empty line

mjauvin committed:
use full path to facade

mjauvin committed:
do not load facade

mjauvin commented:
I need help with this. The tests are working in my local install and with github it says it cannot find Config.
I tried including the facade but no luck.

LukeTowers commented:
@mjauvin try accessing the config through `$this->app->config()`

mjauvin committed:
access config through the app instance

mjauvin committed:
remove empty line

mjauvin commented:
It's really odd, why are the tests running fine locally and not on the github environment?

mjauvin committed:
use setUp() method

mjauvin committed:
remove empty line

mjauvin committed:
more fix

mjauvin committed:
add return type to setUp()

mjauvin committed:
update to latest develop

bennothommo commented:
@mjauvin You cannot use facades in the unit tests because the framework is not technically loaded during them - it instead populates the `$this->app` property with a mock of the application.

You instead need to instantiate the class that the Facade provides and use it directly.

mjauvin committed:
Merge branch 'develop' into path-helpers

mjauvin committed:
remove extra setup crap, it is all handled by the mockup application

mjauvin commented:
@LukeTowers I'm about to throw something through the window...

seanthepottingshed commented:
@mjauvin 

Feel free to throw me out of the window, it is Friday after all!

mjauvin commented on src/Support/helpers.php:
```diff
@@ -198,7 +198,22 @@ function plugins_path($path = '')
+     */
+    function media_path($path = '')
+    {
+        $media_path = rtrim(Config::get('cms.storage.media.path', app('path.media')), '/');
```
@LukeTowers Should we be fetching `Config::get('cms.storage.media.path')` here or would it make more sense in Foundation\Application::mediaPath() ...

LukeTowers replied:
Probably whatever the pattern is for the other helpers

mjauvin committed:
add stub Config class

mjauvin committed:
resolve code quality warning

mjauvin committed:
try app() instead

mjauvin commented:
Whoohooooo!!!

mjauvin committed:
add path suffix tests & config_path() test

mjauvin committed:
replace static directory seperator for windows compatibility

mjauvin committed:
make paths portable

mjauvin commented:
@LukeTowers tests done!

mjauvin committed:
use basePath as default prefix

bennothommo commented on src/Support/helpers.php:
```diff
@@ -172,7 +172,7 @@ function traceSql()
     function config_path($path = '')
     {
-        return app('path.config').($path ? '/'.$path : $path);
+        return app('path.config').($path ? DIRECTORY_SEPARATOR.$path : $path);
```
@mjauvin I wonder if we have to, perhaps, make sure the correct directory separator is also used on the `$path` value provided for this method and the other path helpers as well, perhaps using `str_replace`. While people using Windows might use the correct separator for their own code, there's the potential for plugins to be using this method and not doing the same handling.

bennothommo replied:
Also, needs spaces around concatenations:
```suggestion
        return app('path.config') . ($path ? DIRECTORY_SEPARATOR . $path : $path);
```

mjauvin replied:
Good idea, I'll submit code to do this, thanks!

bennothommo commented on src/Support/helpers.php:
```diff
@@ -185,7 +185,7 @@ function config_path($path = '')
     function plugins_path($path = '')
     {
-        return app('path.plugins').($path ? '/'.$path : $path);
+        return app('path.plugins').($path ? DIRECTORY_SEPARATOR.$path : $path);
```
```suggestion
        return app('path.plugins') . ($path ? DIRECTORY_SEPARATOR . $path : $path);
```

bennothommo commented on src/Support/helpers.php:
```diff
@@ -198,7 +198,22 @@ function plugins_path($path = '')
     {
-        return app('path.uploads').($path ? '/'.$path : $path);
+        $uploads_path = rtrim(Config::get('cms.storage.uploads.path', app('path.uploads')), DIRECTORY_SEPARATOR);
+        return $uploads_path.($path ? DIRECTORY_SEPARATOR.ltrim($path, DIRECTORY_SEPARATOR) : $path);
```
```suggestion
        return $uploads_path . ($path ? DIRECTORY_SEPARATOR . ltrim($path, DIRECTORY_SEPARATOR) : $path);
```

bennothommo replied:
@mjauvin Do you need to do this path trimming on the other helpers too?

mjauvin replied:
@bennothommo Yes, all gonna be handled by the next commit...

bennothommo commented on src/Support/helpers.php:
```diff
@@ -198,7 +198,22 @@ function plugins_path($path = '')
+    function media_path($path = '')
+    {
+        $media_path = rtrim(Config::get('cms.storage.media.path', app('path.media')), DIRECTORY_SEPARATOR);
+        return $media_path.($path ? DIRECTORY_SEPARATOR.ltrim($path, DIRECTORY_SEPARATOR) : $path);
```
```suggestion
        return $media_path . ($path ? DIRECTORY_SEPARATOR . ltrim($path, DIRECTORY_SEPARATOR) : $path);
```

bennothommo commented on src/Support/helpers.php:
```diff
@@ -211,7 +226,7 @@ function uploads_path($path = '')
     function themes_path($path = '')
     {
-        return app('path.themes').($path ? '/'.$path : $path);
+        return app('path.themes').($path ? DIRECTORY_SEPARATOR.$path : $path);
```
```suggestion
        return app('path.themes') . ($path ? DIRECTORY_SEPARATOR . $path : $path);
```

bennothommo commented on src/Support/helpers.php:
```diff
@@ -224,7 +239,7 @@ function themes_path($path = '')
     function temp_path($path = '')
     {
-        return app('path.temp').($path ? '/'.$path : $path);
+        return app('path.temp').($path ? DIRECTORY_SEPARATOR.$path : $path);
```
```suggestion
        return app('path.temp') . ($path ? DIRECTORY_SEPARATOR . $path : $path);
```

bennothommo commented on tests/Foundation/ApplicationTest.php:
```diff
@@ -0,0 +1,32 @@
+
+class ApplicationTest extends TestCase
+{
+    protected function setUp():void
```
```suggestion
    protected function setUp(): void
```

bennothommo commented on tests/Support/HelpersTest.php:
```diff
@@ -0,0 +1,80 @@
+        $path = 'extra-path';
+        $types = ['config', 'temp', 'plugins', 'themes'];
+        foreach ($types as $type) {
+            $method = $type.'_path';
+            $this->assertEquals($method($path), app('path.'.$type).DIRECTORY_SEPARATOR.$path);
```
```suggestion
            $method = $type . '_path';
            $this->assertEquals($method($path), app('path.' . $type) . DIRECTORY_SEPARATOR . $path);
```

bennothommo commented on tests/Support/HelpersTest.php:
```diff
@@ -0,0 +1,80 @@
+        $path = 'extra-path';
+        $types = ['uploads', 'media'];
+        foreach ($types as $type) {
+            $method = $type.'_path';
+            $config = 'cms.storage.'.$type.'.path';
+            $this->assertEquals($method($path), Config::get($config, app('path.'.$type)).DIRECTORY_SEPARATOR.$path);
```
```suggestion
            $method = $type . '_path';
            $config = 'cms.storage.' . $type . '.path';
            $this->assertEquals($method($path), Config::get($config, app('path.' . $type)) . DIRECTORY_SEPARATOR . $path);
```

Commit:
Fix concatenations and other code smells

LukeTowers commented on tests/Foundation/ApplicationTest.php:
```diff
@@ -0,0 +1,32 @@
+        $this->assertEquals($this->app->themesPath(), $this->app->buildPath('/themes'));
+        $this->assertEquals($this->app->tempPath(), $this->app->buildPath('/storage/temp'));
+        $this->assertEquals($this->app->uploadsPath(), $this->app->buildPath('/storage/app/uploads'));
+        $this->assertEquals($this->app->mediaPath(), $this->app->buildPath('/storage/app/media'));
```
I'd like to see testing for the buildPath method, since the other methods are basically small wrappers around it, it's more likely that it would break without breaking these tests.

mjauvin committed:
fix reversed test conditions

mjauvin committed:
add assertion test with extra directory separator

mjauvin committed:
add buildPath() method tests

mjauvin committed:
replace slashes in paths with DIRECTORY_SEPARATOR

mjauvin committed:
also replace slashes with DIRECTORY_SEPARATOR in expected assertion results

mjauvin commented on src/Support/helpers.php:
```diff
@@ -926,6 +964,7 @@ function title_case($value)
      */
     function resolve_path($path)
     {
+        $path = str_replace('/', DIRECTORY_SEPARATOR, $path);
```
@bennothommo I replaced slashes here as well, is it ok?

bennothommo replied:
@mjauvin The PathResolver already normalises the slashes, so it doesn't need this line. I'll change it up with my fixes.

mjauvin committed:
replace slashes with DIRECTORY_SEPARATOR for setter methods as well

mjauvin commented:
I believe this is complete.

bennothommo commented:
@mjauvin one minor change - sorry - would you mind changing the `join_path` helper to a `join` method in the `October\Rain\Filesystem\PathResolver` class? I'm hesitant to add too many helpers because they're easily overwritten and pollute the global namespace.

mjauvin committed:
add join() method to PathResolver class and use it instead of a new global helper method

mjauvin commented:
Done @bennothommo

LukeTowers commented on src/Filesystem/PathResolver.php:
```diff
@@ -114,6 +114,25 @@ public static function within($path, $directory)
     }
 
+    /**
+     * join two paths, making sure they use DIRECTORY_SEPARATOR
```
```suggestion
     * Join two paths, making sure they use DIRECTORY_SEPARATOR
```

Commit:
Update src/Filesystem/PathResolver.php

LukeTowers commented on src/Foundation/Application.php:
```diff
@@ -456,6 +540,24 @@ public function getCachedPackagesPath()
+        $prefix = str_replace('/', DIRECTORY_SEPARATOR, $prefix);
+        $path = str_replace('/', DIRECTORY_SEPARATOR, $path);
+
+        return rtrim($prefix, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR . ltrim($path, DIRECTORY_SEPARATOR);
```
Why not use the new ::join() helper method here instead of duplicating the normalization work?

mjauvin committed:
reuse PathResolver::join() method

mjauvin committed:
Merge branch 'path-helpers' of github.com:mjauvin/library into path-helpers

bennothommo commented:
I'm going to clean this up a little before merging this - I think some of the functionality of the `join` method could be put into the `normalizePath` method, and that method could be adjusted a little as well. Should hopefully have it done today.

mjauvin commented:
@bennothommo you might want to add unit tests for the new join() method as well while you're at it.

bennothommo commented:
@mjauvin will do!

LukeTowers commented on src/Foundation/Application.php:
```diff
@@ -182,7 +208,66 @@ public function setThemesPath($path)
+     */
+    public function setTempPath($path)
+    {
+        $path = str_replace('/', DIRECTORY_SEPARATOR, $path);
```
Is there something in the PathResolver @bennothommo @mjauvin that could replace this? It's a bit of a code smell to have a one-liner path cleaner piece of code copy-pasted across every method instead of using a standard method to handle the common goal of making sure the input will generate clean output or fail appropriately regardless of what that input is.

mjauvin replied:
@LukeTowers I've replaced with @bennothommo 's PathResolver::standardize() method calls.

Commit:
PathResolver improvements

- Add "standardize" method which converts directory separators to OS-required characters.
- Simplify "join" method
- Allow "normalize" method to not prepend CWD if required
- Bit of misc. cleanup

mjauvin committed:
themes and plugins also have path settings in cms config

mjauvin committed:
fix tests

mjauvin committed:
improve tests

mjauvin committed:
find the app instance

mjauvin committed:
let go of themes/plugins config

Commit:
Merge branch 'path-helpers' of github.com:mjauvin/library into pr/mjauvin/493

Commit:
Appease the code sniffer

mjauvin committed:
fix application tests

mjauvin committed:
use PathResolver::standardize()

mjauvin committed:
add custom path config in setUp()

mjauvin committed:
improve code spacing and standardize path methods

mjauvin committed:
do not provide default value for the path config settings

mjauvin committed:
get rid of buildPath in favor of PathResolver::join()

mjauvin committed:
add more unit tests for the path methods in Foundation/Application

mjauvin committed:
build our own storage path for the tests

mjauvin committed:
pass code quality check

mjauvin committed:
pass more code quality check

mjauvin committed:
use config from app object

mjauvin committed:
try to get the application object

mjauvin committed:
iteration #238

mjauvin committed:
set config within test function

mjauvin committed:
reuse stub config class

Commit:
Merge remote-tracking branch 'origin/develop' into pr/mjauvin/493

Commit:
Include Mockery library to allow facade mocks

Commit:
Set temp path for stub application, just to be safe

Commit:
Correctly stub application and Config facade for helper tests

bennothommo commented:
@mjauvin So it appears the best way to mock facades is to use the Mockery library which Laravel uses. I've ended up just fixing the tests fully this way. If you want to see how it's done, have a look starting from here: https://github.com/octobercms/library/pull/493/files#diff-46ffd9a10d01a153dc357c4ac0c5d599e13c4d95a90a6d2c4e81bdce551b4318R11

LukeTowers commented on src/Support/helpers.php:
```diff
@@ -165,14 +165,15 @@ function traceSql()
     {
-        return app('path.config').($path ? '/'.$path : $path);
+        $config_path = app('path.config');
+        return PathResolver::join($config_path, $path);
```
Why can't these functions be one-liners?

Commit:
Convert path helpers to one-liners

mjauvin committed:
remove unused config

mjauvin commented:
@LukeTowers @bennothommo I don't see the point in converting the path helpers to one-liners, at least not for `uploads_path()` and `media_path()`

mjauvin commented:
Otherwise, I'd say it's looking pretty good now, what do you guys think?

mjauvin committed:
cleanup

LukeTowers commented:
@mjauvin yup, looking pretty good. What docs changes are required, and how would you describe these changes in the changelog?

mjauvin commented:
@LukeTowers https://github.com/octobercms/docs/pull/490

Changelog should only be:
- fixes uploads_path() broken helper method
- add media_path() helper method

mjauvin commented:
For reference, the plugins_path() helper method was doing this:
```
return app('path.plugins').($path ? '/'.$path : $path);
```
But `app('path.plugins')` did not exist.

LukeTowers commented:
How was it working previously then? Because I used plugins_path extensively in my own projects and I don't think I've ever had a problem with it?

mjauvin commented:
@LukeTowers Sorry, I meant to write "uploads_path()" method was broken... my bad.

mjauvin commented:
@LukeTowers @bennothommo Aren't the provided tests enough for "testing needed" ?
